### ISSUE TITLE
Simplify `SetFormula`

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#9457]
- [x] Remove `NNF` 
- [x] Remove `propagation`
- [x] Rework `Union` to use `TwoList`
- [ ] Rework `Intersec` to use `TwoList`
- [ ] Remove `mapSubformulas`
- [x] Remove `tableForm`
- [x] Remove `variables`
- [x] Remove or replace `isEquivalent` by something simpler 
  - Zhegalkin gives this check for free, so it seems expensive to do it upfront at a cost
  - We could still have a cheap syntactic check though.

**Comment by magnus-madsen** [comment]
@JonathanStarup Ideas for simplifications

**Comment by JonathanStarup** [comment]
> Zhegalkin gives this check for free, so it seems expensive to do it upfront at a cost

after a not-free conversion (i.e. checking a union b union c is probably cheaper than converting to zhegalkin and checking)

**Comment by magnus-madsen** [comment]
> > Zhegalkin gives this check for free, so it seems expensive to do it upfront at a cost
> 
> after a not-free conversion (i.e. checking a union b union c is probably cheaper than converting to zhegalkin and checking)

Is the call to evaluate etc also not expensive?

**Comment by magnus-madsen** [comment]
We can benchmark it.

**Comment by JonathanStarup** [comment]
yeah it might be faster, i justed wanted to put an asterisk on free

**Comment by magnus-madsen** [comment]
How is that possible? its very significant (this was my computer though)

**Comment by JonathanStarup** [comment]
I think we should also start to think about how to present types to the user - that will require minimization that can't rely on ANF

**Comment by magnus-madsen** [comment]
> I think we should also start to think about how to present types to the user - that will require minimization that can't rely on ANF

Yes, but I think it should be a new class 100% separate from the compiler. Purely at "PrettyPrintEff" (or whatever). Then its complexity can be as bad as we need.

**Comment by JonathanStarup** [comment]
> How is that possible? its very significant (this was my computer though)

what did you change? yeah seems a lot

**Comment by magnus-madsen** [comment]
I removed this code:

```
      // f1 ~ f2, where f1 and f2 are ground
      // ---
      // {}, [] if solved
      // {f1 ~error f2}, [] if unsolvable
      case (f1, f2) if f1.isGround && f2.isGround =>
        if (isEquivalent(f1, f2)) {
          Some(Nil, SetSubstitution.empty)
        } else {
          Some(List(eq.toUnsolvable), SetSubstitution.empty)
        }
```

**Comment by JonathanStarup** [comment]
> > I think we should also start to think about how to present types to the user - that will require minimization that can't rely on ANF
> 
> Yes, but I think it should be a new class 100% separate from the compiler. Purely at "PrettyPrintEff" (or whatever). Then its complexity can be as bad as we need.

yeah i think its gonna be pretty bad

**Comment by mlutze** [comment]
Is this still relevant?

**Comment by magnus-madsen** [comment]
Not really, we need to rework SetFormula to use Zhegalkin instead.
